### PR TITLE
Adjust to zss1.2.1release

### PR DESCRIPTION
**Pull request by cdjackson**
Replaces #442 

@triller-telekom as discussed on Friday. Note that I've updated to use the latest SNAPSHOT of the libraries. This incorporates the changes to resolve (hopefully!) the node added/updated notifications.

It also updates for the data store, and I changed a few things in the data store in the libs this morning to improve performance slightly.

I suggest to test this out, and if there are no further issues in the coming days we merge.

Note that I've not been able to test this until the IDE issues are resolved.

Also-by: Henning Sudbrock henning.sudbrock@telekom.de
Also-by: Stefan Triller stefan.triller@telekom.de
Signed-off-by: Chris Jackson chris@cd-jackson.com

### CODE REVIEW

File: org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/ZigBeeDataStore.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        stream.alias("PowerSourceType", PowerSourceType.class);
+        stream.alias("FrequencyBandType", FrequencyBandType.class);
+        stream.alias("BindingTable", BindingTable.class);
+        stream.alias("IeeeAddress", BindingTable.class);
+        stream.registerConverter(new IeeeAddressConverter());
+        return stream;
+    }
+
+    private File getFile(IeeeAddress address) {
+        return new File(networkStateFilePath + address + ".xml");
```
Comment (triller-telekom) [inline org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/internal/ZigBeeDataStore.java]: There is a `File.separator` missing between the `networkStateFilePath` and the `address`

**Comment by triller-telekom** [comment]
Apart from the one inline comment where I found a small bug, the stuff is working.

I have cherry-picked your commits into our build environment and have successfully tested it in on our SPI ember device and with an ember usb and telegesis usb stick.

**Comment by cdjackson** [comment]
Thanks Stefan - and is that also using the 1.2.1-SNAPSHOT of the libraries?

**Comment by triller-telekom** [comment]
Since I do not see a 1.2.1 tag in ZSS, I don't know. But I am using the commit https://github.com/zsmartsystems/com.zsmartsystems.zigbee/commit/58ae488c4612a4977227b07920466838647a6542 from ZSS with this PR and it worked.

**Comment by triller-telekom** [comment]
> You should be able to use the 1.2.1-SNAPSHOT of the libraries using Maven

We are in a sort of build environment hell and I cannot just replace the version string in a maven artifact. I am building against our own artifacts.

Could you please point me to the commits you are referring to which should be in my ZSS version, on top of the commit I just mentioned?

**Comment by cdjackson** [comment]
I think that this commit is the latest -:

https://github.com/zsmartsystems/com.zsmartsystems.zigbee/pull/671

I will likely merge this later today if tests by other confirm it's all working (initial feedback is positive) and I think this will then become 1.2.1, and we should then merge that into OH binding.

**Comment by cdjackson** [comment]
This seems to be working ok for me, and also tested by @triller-telekom so merging (finally!).

**Comment by wborn** [comment]
Are these 1.2.1 zsmartsystems artifacts released versions? It looks like they are only available via the repo that got added in this PR (https://oss.jfrog.org/artifactory/oss-snapshot-local). Because openhab-distro doesn't seem to use this repo the [openHAB-Distribution](https://ci.openhab.org/job/openHAB-Distribution) now fails:

`[ERROR] Message: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=openhab-binding-zigbee; type=karaf.feature; version=2.5.0.SNAPSHOT; filter:="(&(osgi.identity=openhab-binding-zigbee)(type=karaf.feature)(version>=2.5.0.SNAPSHOT))" [caused by: Unable to resolve openhab-binding-zigbee/2.5.0.SNAPSHOT: missing requirement [openhab-binding-zigbee/2.5.0.SNAPSHOT] osgi.identity; osgi.identity=org.openhab.binding.zigbee.cc2531; type=osgi.bundle; version="[2.5.0.201907290432,2.5.0.201907290432]"; resolution:=mandatory [caused by: Unable to resolve org.openhab.binding.zigbee.cc2531/2.5.0.201907290432: missing requirement [org.openhab.binding.zigbee.cc2531/2.5.0.201907290432] osgi.wiring.package; filter:="(osgi.wiring.package=org.openhab.binding.zigbee.handler)" [caused by: Unable to resolve org.openhab.binding.zigbee/2.5.0.201907290432: missing requirement [org.openhab.binding.zigbee/2.5.0.201907290432] osgi.wiring.package; filter:="(&(osgi.wiring.package=com.zsmartsystems.zigbee)(version>=1.2.0)(!(version>=2.0.0)))"]]]`

CC: @openhab/distro-maintainers

**Comment by cdjackson** [comment]
Yes, they are released versions. SNAPSHOT versions only are in the oss-snapshot repository.

https://dl.bintray.com/zsmartsystems/com.zsmartsystems/com/zsmartsystems/zigbee/1.2.1/

This was passing CI yesterday.

**Comment by kaikreuzer** [comment]
https://dl.bintray.com/zsmartsystems/com.zsmartsystems/com/zsmartsystems/zigbee/1.2.1/ is not an openHAB repo. Could you release them to Maven Central?

**Comment by kaikreuzer** [comment]
The bintray repo is currently added to the virtual repo https://openhab.jfrog.io/openhab/libs-release.
But checking https://dl.bintray.com/zsmartsystems/com.zsmartsystems/com/zsmartsystems/zigbee/1.2.1/, I can only see the pom.xml there and nothing else. @cdjackson Are you sure that the deployment was successful?

**Comment by cdjackson** [comment]
I thought that the repository was defined in the pom, but I will check tonight. Does it really need to be released into an "openHAB repository" if it's defined in the POM?

CI seemed to resolve this ok yesterday before I merged it.

**Comment by kaikreuzer** [comment]
> if it's defined in the POM?

It should not be defined in a pom - we only use repos that are defined in the central super pom and should not have any exceptions to this. So ideally, 3rd party dependencies should be available in Maven Central.

**Comment by cdjackson** [comment]
I'm pretty sure that the CI yesterday was ok unless I misunderstand the logs (which is certainly possible ;) ) -:
https://ci.openhab.org/view/Pull%20Request%20Builds/job/PR-org.openhab.binding.zigbee/635/consoleFull

image

image

**Comment by kaikreuzer** [comment]
Well, can you see the artifacts in your bintray repo? I can't...

**Comment by kaikreuzer** [comment]
I am not saying they weren't there yesterday, I only see that they are not there today.

**Comment by cdjackson** [comment]
I have not looked, but I don't know why they would dissappear. I've used this repo for a couple of years.

Let me check in an hour or so when I have some time.

**Comment by kaikreuzer** [comment]
Just click https://dl.bintray.com/zsmartsystems/com.zsmartsystems/com/zsmartsystems/zigbee/1.2.1/ and you should immediately see it (or rather not) yourself.

**Comment by cdjackson** [comment]
But that's just the pom location isn't it? The data is stored elsewhere and it is there for all the different versions. I've never had reason to question if this is right or not - its what bintray always does, and I've never had issues resolving these dependencies with other users before.

eg - all the com.zsmartsystems.zigbee artifacts are here -:

https://dl.bintray.com/zsmartsystems/com.zsmartsystems/com/zsmartsystems/zigbee/com.zsmartsystems.zigbee/1.2.1/

If I run this check through a different system, it also finds the artifacts so I think this is correct -:

https://mvnrepository.com/artifact/com.zsmartsystems.zigbee/com.zsmartsystems.zigbee/1.2.1

**Comment by kaikreuzer** [comment]
Ok, sorry, you are right.
I can indeed also see the artifacts in our JFrog repo at https://openhab.jfrog.io/openhab/libs-release/com/zsmartsystems/zigbee/com.zsmartsystems.zigbee/1.2.1/.

So let's assume it was just some hiccup and not-yet-synched issue - I'll restart the distro build and see if it still occurs.

**Comment by kaikreuzer** [comment]
It didn't help. :-(

I see in the log of Jenkins:
``´
[ERROR] Resources: {
[ERROR] mvn:com.eclipsesource.jaxrs/publisher/5.3.1
[ERROR] mvn:com.google.guava/guava/18.0
[ERROR] mvn:com.google.guava/guava/21.0
[ERROR] mvn:com.zsmartsystems.zigbee/com.zsmartsystems.zigbee.dongle.cc2531/1.1.6
[ERROR] mvn:com.zsmartsystems.zigbee/com.zsmartsystems.zigbee.dongle.ember/1.1.6
[ERROR] mvn:com.zsmartsystems.zigbee/com.zsmartsystems.zigbee.dongle.telegesis/1.1.6
[ERROR] mvn:com.zsmartsystems.zigbee/com.zsmartsystems.zigbee.dongle.xbee/1.1.6
[ERROR] mvn:com.zsmartsystems.zigbee/com.zsmartsystems.zigbee/1.1.6
```
It seems as if it still builds against 1.1.6, so then it is plausible that it cannot find version 1.2.1.

**Comment by kaikreuzer** [comment]
I see that 1.1.6 is hardcoded in https://github.com/openhab/openhab-distro/blob/master/features/addons/src/main/feature/feature.xml#L9-L13, which does not make much sense.
Imho, we should move the Karaf feature for the Zigbee binding to the zigbee repo, wdyt?

**Comment by kaikreuzer** [comment]
I've created https://github.com/openhab/openhab-distro/pull/949 as a quick fix.

**Comment by wborn** [comment]
> we should move the Karaf feature for the Zigbee binding to the zigbee repo, wdyt?

That makes a lot of sense. It may prevent such issues in the future and it's easier to update the dependencies/versions in a single PR.

**Pull request merged**